Implement Tar async APIs

carlossanlop opened the pull request:
Addresses: https://github.com/dotnet/runtime/issues/65951

Submitting it as a draft for these reasons:

- [x] This PR depends on another that needs to be merged first: https://github.com/dotnet/runtime/pull/70325
- [X] I want to first show each individual async method implemented identically to the synchronous one, with the asynchronous differences only.
- [x] Add unit tests.
- [x] Public documentation.
- [x] Remove duplicate code between sync and async methods.
- [ ] Address feedback.

carlossanlop committed:
ref: TarEntry.ExtractToFileAsync

carlossanlop committed:
src: Implement TarEntry.ExtractToFileAsync

carlossanlop committed:
ref: Add TarReader.GetNextEntryAsync

carlossanlop committed:
ref: Add TarWriter.DisposeAsync

carlossanlop committed:
src: Implement TarWriter.DisposeAsync

carlossanlop committed:
ref: Add TarReader.DisposeAsync

carlossanlop committed:
src: Implement TarReader.DisposeAsync

stephentoub commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:
```diff
@@ -217,29 +219,33 @@ public void ExtractToFile(string destinationFileName, bool overwrite)
+            if (EntryType is TarEntryType.SymbolicLink or TarEntryType.HardLink)
+            {
+                throw new InvalidOperationException(string.Format(SR.TarEntryTypeNotSupportedForExtracting, EntryType));
+            }
```
Typically any exceptions other than argument exceptions are wrapped in the returned task, e.g.
```C#
return Task.FromException(new InvalidOperationException(string.Format(SR.TarEntryTypeNotSupportedForExtracting, EntryType));
```

carlossanlop replied:
Will do.

Why do we not wrap argument exceptions?

carlossanlop replied:
This feels like a good candidate for a Roslyn analyzer, if we don't have one yet.

stephentoub replied:
> Why do we not wrap argument exceptions?

Because in typical usage they represent a 100% deterministic developer error in how the API is being used.  We don't want such errors being eaten by the returned task, especially if doing so could lead to a deadlock, e.g. if this task won't end up performing some work that would ultimately unblock something else.

> And what about ArgumentNullException?

That is an argument exception.  So same deal.

carlossanlop replied:
Thanks for the explanation.

I opened this API suggestion in case it makes sense to add such analyzer: https://github.com/dotnet/runtime/issues/70905

carlossanlop replied:
@stephentoub does it make sense to add async methods all the way down to Interop.IOErrors.cs? I noticed we throw exceptions there, but on async code they would not be thrown properly as described above.

I'm thinking of adding methods like this one:

```cs
    internal static Task<int> CheckIoAsync(int result, string? path = null, bool isDirectory = false, Func<ErrorInfo, ErrorInfo>? errorRewriter = null) =>
        result < 0 ?
            Task.FromException<int>(GetExceptionForIoErrno(Sys.GetLastErrorInfo(), path, isDirectory, errorRewriter)) :
            Task.FromResult(result);
```

stephentoub replied:
Can you give me an example of what wouldn't be handled properly?

carlossanlop replied:
I have this method in TarEntry.Unix.cs:
```cs
        // Unix specific implementation of the method that extracts the current entry as a hard link.
        private void ExtractAsHardLink(string targetFilePath, string hardLinkFilePath)
        {
            Debug.Assert(EntryType is TarEntryType.HardLink);
            Debug.Assert(!string.IsNullOrEmpty(targetFilePath));
            Debug.Assert(!string.IsNullOrEmpty(hardLinkFilePath));
            Interop.CheckIo(Interop.Sys.Link(targetFilePath, hardLinkFilePath), hardLinkFilePath);
        }
```

And I want to write an appropriate async version of this method. The `Interop.CheckIo` method is supposed to throw if the p/invoke fails. So I was thinking of adding a method like the one I shared in my previous comment, and then call it like this in the new async method:

```cs
        // Unix specific implementation of the method that asynchronously extracts the current entry as a hard link.
        private Task ExtractAsHardLinkAsync(string targetFilePath, string hardLinkFilePath)
        {
            Debug.Assert(EntryType is TarEntryType.HardLink);
            Debug.Assert(!string.IsNullOrEmpty(targetFilePath));
            Debug.Assert(!string.IsNullOrEmpty(hardLinkFilePath));
            return Interop.CheckIoAsync(Interop.Sys.Link(targetFilePath, hardLinkFilePath), hardLinkFilePath);
        }
```

Are there scenarios where I don't need to wrap the exception with `[Value]Task.FromException`?

carlossanlop replied:
I know there isn't one, but my question was around handling the exception.

stephentoub replied:
There's no reason to create asynchronous version of ExtractAsHardLinkAsync or CheckIo, so I'm not clear on your question. Can you clarify?

carlossanlop replied:
So the async method that calls `ExtractAsHardLink` should wrap that in a try catch, and the catch should return `Task.FromException`?

```cs
internal Task MethodAsync(...)
{
try
{
    ExtractAsHardLink(...);
    return Task.CompletedTask;
}
catch (Exception e)
{
    return Task.FromException(e);
}
}
```

carlossanlop replied:
In other words, I should only create a new async method if it will call another async method inside it.

But in the case of methods that can exceptions, the last async method in the callstack is the one which should handle the exception by catching it and return it inside a Task.

Is this correct?

bartonjs replied:
What really matters is the public function black-box experience.

In the call of

```C#
Task t = someObjectOrType.FooAsync(arg0, arg1, arg2);
t.Wait();
Console.WriteLine(t.Exception);
```

There are two possibilities for the exception:

1) the call to FooAsync was invalid, in which case the FooAsync method itself should throw, and we never make it to line 2.  (ArgumentException)
2) We created a Task. The task ran... until it failed.  The exception is now in the t.Exception property.

For the latter, this is accomplished in one of three ways.

I) After argument validation, FooAsync failed synchronously, and the implementation returns `Task.FromException(e)`.
II) You're building tasks by chaining them together with Task.ContinueWith(action), inside the action you throw an exception (as per normal).  The Task system records the exception and changes the task state to Faulted.  https://github.com/dotnet/runtime/blob/988f570dc4ddd79a166ca8cbc184b2938ed2fefc/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2332-L2357
III) You're writing a method with `async`/`await`.  That's just a fancy way of doing II.

What that boils down to is:

* In the public method you might need to be aware of being "inside" or "outside" the task with an exception, here's where Task.FromException comes in.
* Once you enter an `async`/`await` method you can stop thinking, just write throw as normal.
* If you're making Tasks by hand, and you're inside a callback, throw as normal.
* If you're making Tasks by hand and you're not inside a callback, Task.FromException may again matter.
  * But why are you making Tasks by hand after the public entrypoint?

stephentoub commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -69,6 +71,59 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
+        internal async ValueTask<bool> TryGetNextHeaderAsync(Stream archiveStream, bool copyData, CancellationToken cancellationToken)
+        {
+            // The four supported formats have a header that fits in the default record size
+            IMemoryOwner<byte> rented = MemoryPool<byte>.Shared.Rent(minBufferSize: TarHelpers.RecordSize);
```
This should use ArrayPool directly rather than MemoryPool.

carlossanlop replied:
But a couple of lines below I need to pass the `Memory<byte>` to `archiveStream.ReadExactlyAsync`.

bartonjs replied:
Sure... but arrays are implicitly `Memory<byte>` (or if you want to slice it, then explicitly use AsMemory)

```
byte[] rented = ArrayPool<byte>.Shared.Rent(TarHelpers.RecordSize);
Memory<byte> buffer = rented.AsMemory(0, TarHelpers.RecordSize);

...

ArrayPool<byte>.Shared.Return(rented);
```

carlossanlop replied:
TIL. Thanks.

carlossanlop replied:
> Sure... but arrays are implicitly `Memory<byte>` (or if you want to slice it, then explicitly use AsMemory)

Thanks for this explanation in advance. I am starting to see some test failures when attempting to do async writes using the existing code that takes spans, like this:

https://github.com/dotnet/runtime/blob/70e9ca05739971856f17dfd23f44bd72e153cc2d/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs#L207-L212

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs:
```diff
@@ -126,16 +128,36 @@ public void WriteEntry(string fileName, string? entryName)
+        /// <exception cref="ObjectDisposedException">The archive stream is disposed.</exception>
+        /// <exception cref="ArgumentException"><paramref name="fileName"/> or <paramref name="entryName"/> is <see langword="null"/> or empty.</exception>
+        /// <exception cref="IOException">An I/O problem occurred.</exception>
+        public async Task WriteEntryAsync(string fileName, string? entryName, CancellationToken cancellationToken = default)
```
@stephentoub this public API returns a `Task`. My immediate assumption was that all the internal async APIs this method calls should also return a `Task`. Is there anything preventing me from changing all the internal methods to return a `ValueTask` instead?

stephentoub replied:
> Is there anything preventing me from changing all the internal methods to return a ValueTask instead?

No, but just making that change will typically only yield benefits in the case of `Task<T>` to `ValueTask<T>`.

bartonjs commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs:
```diff
@@ -117,16 +129,36 @@ public void WriteEntry(string fileName, string? entryName)
+        {
+            ThrowIfDisposed();
+
+            ArgumentException.ThrowIfNullOrEmpty(fileName);
```
The ArgumentException, and possibly the ObjectDisposedException should generally be "outside" the Task.  So this method shouldn't be `public async Task`, just `public Task`, do argument validation, then it can call an `async Task` private or local member.

carlossanlop committed:
Don't use partial methods in TarEntry.

carlossanlop committed:
Don't use partial methods in TarWriter.

bartonjs commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -69,6 +71,59 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
+            }
+            finally
+            {
+                rented.Dispose();
```
(Which, I guess, means you can get rid of the whole try/finally)

stephentoub replied:
> Async methods shouldn't return to the pools via a finally

Why?

bartonjs replied:
I'm mainly parroting Levi here for "don't return to a pool from a finally in async code", but the reasoning had something to do with the possibility of a native async operation still doing writes when the managed wrapper task has thrown a timeout exception, so returning to the pool in that case could result in the async op writing to the buffer after it has been rented elsewhere.

I remember reading a guidance document at some point, but I can't find it offhand.

From synchronous code, the expectation is more that there shouldn't be a possibility of a second reference, so returning from a finally is OK (though my recollection was that the guidance suggested not to, for a uniformity of style).

stephentoub replied:
> but the reasoning had something to do with the possibility of a native async operation still doing writes when the managed wrapper task has thrown a timeout exception, so returning to the pool in that case could result in the async op writing to the buffer after it has been rented elsewhere.

That would be a bug in such an async method.  The task returned from an async method should not complete until all work associated with the operation has quiesced, all resources are no longer used, etc.

> From synchronous code, the expectation is more that there shouldn't be a possibility of a second reference, so returning from a finally is OK (though my recollection was that the guidance suggested not to, for a uniformity of style).

The exact same issue could exist with synchronous code, if the implementation similarly erroneously returned/threw while resources were still in use.

carlossanlop committed:
TarFile exceptions async handling.

carlossanlop committed:
Move TarWriter awaited code to separate methods.

carlossanlop committed:
src: Implement TarReader.GetNextEntryAsync

carlossanlop committed:
ref: Add TarWriter.WriteEntryAsync

carlossanlop committed:
src: Implement TarWriter.WriteEntryAsync

carlossanlop committed:
ref: Add TarFile.CreateFromDirectoryAsync

carlossanlop committed:
src: Implement TarFile.CreateFromDirectoryAsync

carlossanlop committed:
ref: Add TarFile.ExtractToDirectoryAsync

carlossanlop committed:
src: Implement TarFile.ExtractToDirectoryAsync

carlossanlop committed:
Update Global Extended Attributes async methods to the latest rebase.

carlossanlop committed:
Remove unused parameter in SetModeOnFile after latest rebase that uses File.SetUnixFileMode.

carlossanlop committed:
Adjust TarEntry async methods to handle exceptions properly.

carlossanlop committed:
ref: Restore System.IAsyncDisposable inheritance in TarReader, it wasn't properly merged with latest rebase.

carlossanlop commented:
As I'm writing my async unit tests, I noticed we do not have CA2007 enabled for test projects ("Consider calling ConfigureAwait on the awaited task"). Is there a reason for that? Should I still add `ConfigureAwait(false)` to all my async calls in tests?

stephentoub commented:
> Is there a reason for that? Should I still add ConfigureAwait(false) to all my async calls in tests?

In general you want the equivalent of ConfigureAwait(true) in tests, which is the default.

carlossanlop committed:
Add async tests for TarReader.TarEntry.ExtractToFile.
Rename TarReader.ExtractToFile* test files to TarReader.TarEntry.ExtractToFile*, since the actual async method belongs to TarEntry. These test methods extract files coming from assets that are opened as archives by a TarReader.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:
```diff
@@ -488,25 +578,45 @@ private void ExtractAsRegularFile(string destinationFileName)
+                }
+                catch (Exception e)
+                {
+                    await Task.FromException(e).ConfigureAwait(false);
```
This is wrong, isn't it?

I'm just reading the comment by @bartonjs here: https://github.com/dotnet/runtime/issues/70905#issuecomment-1159117124

So I assume I need to catch exceptions in the calling method, not here?

stephentoub replied:
I don't understand what this is trying to do.  As written it's an expensive nop that eats all exceptions thrown.

This is all inside of an async method.  An async method catches all exceptions that go unhandled in the body and stores them into the Task being returned.  Which is what you want here... just remove the try/catch and let any exceptions propagate.

I think you're making it harder than it actually is.  In general, let exceptions propagate exactly as you normally would in a synchronous method.  The only special-cases typically are when doing argument validation in the entry point method.

carlossanlop replied:
> I think you're making it harder than it actually is

I know, that's why I'm asking 🙂. There are many things about async code that I am unfamiliar with, so I appreciate the patience and the explanations.

> As written it's an expensive nop that eats all exceptions thrown.

Ah alright, so I had the wrong understanding.

> In general, let exceptions propagate exactly as you normally would in a synchronous method. The only special-cases typically are when doing argument validation in the entry point method.

Thanks, that is the answer I was looking for. I'll adjust the code based on this.

carlossanlop commented:
I still need to add checks for cancelled CancellationTokens. I'm looking at examples in our code base.

carlossanlop committed:
Adjust rented buffer handling based on feedback.

carlossanlop committed:
This needs to be awaited.

carlossanlop committed:
Reserve ArrayPool in TarWriter WriteEntryAsyncInternal, not MemoryPool. Remove the try finally, there is no return value in this method.

carlossanlop committed:
Split writing methods in TarHeader into span based and memory based.

carlossanlop committed:
tests: TarReader.GetNextEntryAsync.

carlossanlop committed:
Uncomment InlineData from sync file reading method.

carlossanlop committed:
Simplify exception handling. Only return *Task.FromException in public entry-point methods. The rest just let them throw.

carlossanlop committed:
Make the TarHeader async methods static if they depend on the struct itself.

carlossanlop committed:
Add missing GEA changes from the last PR.

carlossanlop committed:
TarReader GetNextEntryAsync tests for file assets.

carlossanlop committed:
Add async compressed tar tests.

carlossanlop committed:
Adjust checksum calculation and saving after writing in async methods.

carlossanlop committed:
Add async V7 checksum verification test.

carlossanlop committed:
Fix await using for FileStreams in TarFile. Move awaited code to internal methods.

carlossanlop committed:
Add async tests for TarFile.CreateFromDirectoryAsync for file assets.

carlossanlop committed:
Add async tests for TarFile.CreateFromDirectoryAsync for streams.

carlossanlop committed:
Add async tests for TarFile.ExtractToDirectoryAsync for file assets.

carlossanlop committed:
Add async tests for TarFile.ExtractToDirectoryAsync for streams.

carlossanlop committed:
Add tests for TarFile.ExtractToDirectoryAsync platform specific.

carlossanlop committed:
Add TarEntry.ExtractToFileAsync tests

carlossanlop committed:
Join "regular entry vs writer format" tests into single theory test.

carlossanlop committed:
Missing using directive in Unix test file.

carlossanlop committed:
Add TarWriter.WriteEntryAsync tests for entries.

carlossanlop committed:
Add TarWriter.WriteEntryAsync tests for file assets.

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:
```diff
@@ -18,8 +19,8 @@ public abstract partial class TarEntry
 
-        // Constructor called when reading a TarEntry from a TarReader.
-        internal TarEntry(TarHeader header, TarReader readerOfOrigin, TarEntryFormat format)
+    // Constructor called when reading a TarEntry from a TarReader.
+    internal TarEntry(TarHeader header, TarReader readerOfOrigin, TarEntryFormat format)
```
nit: indentation 
```suggestion
        // Constructor called when reading a TarEntry from a TarReader.
        internal TarEntry(TarHeader header, TarReader readerOfOrigin, TarEntryFormat format)
```

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -63,6 +90,25 @@ internal static void CopyBytes(Stream origin, Stream destination, long bytesToCo
         }
 
+        // Asynchronously helps copy a specific number of bytes from one stream into another.
+        internal static async ValueTask CopyBytesAsync(Stream origin, Stream destination, long bytesToCopy, CancellationToken cancellationToken)
```
note to self: it's a great use case for `SubStream` (we don't have it yet):

```cs
=> origin.Slice(0, bytesToCopy).CopyToAsync(destination, cancellationToken);
```

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -63,6 +90,25 @@ internal static void CopyBytes(Stream origin, Stream destination, long bytesToCo
+                int currentLengthToRead = (int)Math.Min(MaxBufferLength, bytesToCopy);
+                Memory<byte> memory = buffer.AsMemory(0, currentLengthToRead);
+                int bytesRead = await origin.ReadAsync(memory, cancellationToken).ConfigureAwait(false);
+                if (bytesRead != currentLengthToRead)
+                {
+                    throw new EndOfStreamException();
+                }
+                await destination.WriteAsync(memory, cancellationToken).ConfigureAwait(false);
+                bytesToCopy -= currentLengthToRead;
```
```suggestion
                if (bytesRead == 0)
                {
                    throw new EndOfStreamException();
                }
                await destination.WriteAsync(memory.Slice(0, bytesRead), cancellationToken).ConfigureAwait(false);
                bytesToCopy -= bytesRead;
```

carlossanlop replied:
Yet another place to consume @eerhardt 's good ole ReadExactly* APIs.

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Windows.cs:
```diff
@@ -77,5 +79,71 @@ public sealed partial class TarWriter : IDisposable
+                    Mode = FileMode.Open,
+                    Access = FileAccess.Read,
+                    Share = FileShare.Read,
+                    Options = FileOptions.None
```
```suggestion
                    Options = FileOptions.Asynchronous
```

carlossanlop committed:
Fix test faiest failure - Windows test accidentally being run on Linux

carlossanlop committed:
Set Options to Asynchornous in TarWriter internal method in Windows. For Unix, add missing FileStreamOptions.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -46,6 +49,30 @@ internal static void AdvanceStream(Stream archiveStream, long bytesToDiscard)
+            }
+            else if (bytesToDiscard > 0)
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(minimumLength: MaxBufferLength);
```
`bytesToDiscard` is a `long`, `MaxBufferLength` is `int`.  So I'll do this instead:

```cs
                int minimumLength = bytesToDiscard <= int.MaxValue ? Math.Min(MaxBufferLength, (int)bytesToDiscard) : MaxBufferLength;
                byte[] buffer = ArrayPool<byte>.Shared.Rent(minimumLength);
```

carlossanlop replied:
Ah no, I can keep using your suggestion, just need to cast the final result (which would always be smaller than `int.MaxValue`):

```cs
byte[] buffer = ArrayPool<byte>.Shared.Rent(minimumLength: (int)Math.Min(MaxBufferLength, bytesToDiscard));
```

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -46,6 +49,30 @@ internal static void AdvanceStream(Stream archiveStream, long bytesToDiscard)
+                {
+                    int currentLengthToRead = (int)Math.Min(MaxBufferLength, bytesToDiscard);
+                    int bytesRead = await archiveStream.ReadAsync(buffer.AsMemory(0, currentLengthToRead), cancellationToken).ConfigureAwait(false);
+                    if (bytesRead != currentLengthToRead)
+                    {
+                        throw new EndOfStreamException();
+                    }
+                    bytesToDiscard -= currentLengthToRead;
```
Actually, this can be changed to use the stream APIs @eerhardt added recently: https://github.com/dotnet/runtime/pull/69272/

Although I will have to keep the loop, since `bytesToDiscard` is a `long`.

carlossanlop committed:
Simplify AdvanceStream* helper methods with Math.Min and ReadExactly*.

carlossanlop committed:
Simplify CopyBytes* helper methods with ReadExactly*.

carlossanlop committed:
Add CancellationToken handling.

carlossanlop committed:
Weird spacing.

carlossanlop committed:
Move awaited code from GetNextEntryAsync to a private method. This prevents throwing OperationCanceledException instead of the expected TaskCanceledException.

carlossanlop committed:
Add tests for throwing with a cancelled CancellationToken

carlossanlop committed:
Simplify async calls in tests.

carlossanlop committed:
Use await using in FileStream used when extracting as regular file.

carlossanlop committed:
Reuse code in TarEntry.ExtractRelativeToDirectory* methods

carlossanlop committed:
Reuse code in TarFile methods.

carlossanlop committed:
Reuse TarWriter.WriteEntry*(string,string) code.

carlossanlop committed:
Add tests for null check in TarWriter.WriteEntry(TarEntry)

carlossanlop committed:
Revert renaming of TarHeader.Read methods. Not needed.

carlossanlop committed:
Remove unnecessary try-finally in TarHeader.TryGetNextHeader

carlossanlop committed:
GetDataStreamAsync should return a null if size is 0, not a MemoryStream.

carlossanlop committed:
Reuse code in TarHeader.ReadExtendedAttributesBlock*

carlossanlop commented:
/azp run runtime-extra-platforms

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:
```diff
@@ -241,25 +363,112 @@ private static void CreateFromDirectoryInternal(string sourceDirectoryName, Stre
 
+        // Determines what should be the base path for all the entries when creating an archive.
+        private static string GetBasePathForCreateFromDirectory(DirectoryInfo di, bool includeBaseDirectory)
+        {
+            string basePath = di.FullName;
+
+            if (includeBaseDirectory && di.Parent != null)
+            {
+                basePath = di.Parent.FullName;
+            }
+
+            return basePath;
+        }
```
it can be one liner?

```suggestion
            => includeBaseDirectory && di.Parent != null ? di.Parent.FullName : di.FullName;
```

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -28,45 +30,162 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
+
+            ProcessDataBlock(archiveStream, copyData);
+
+            ArrayPool<byte>.Shared.Return(rented);
```
since `RecordSize` is quite small (512 bytes), there is no need to use `ArrayPool`, we can just stack allocate the buffer

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -28,45 +30,162 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
+                case TarEntryType.TapeVolume: // Might contain data
+                default: // Unrecognized entry types could potentially have a data section
+                    header._dataStream = await GetDataStreamAsync(archiveStream, copyData, header._size, cancellationToken).ConfigureAwait(false);
+                    if (header._dataStream is SeekableSubReadStream)
```
do we need to check the type?

```suggestion
                    if (header._dataStream.CanSeek)
```

carlossanlop replied:
Checking if it is SeekableSubReadStream also saves a nullcheck.

But also yes, it's necessary to check if the data stream is one of those specific two types because the GetDataStream method can also return a MemoryStream, when the underlying archive is unseekable and the user chooses to copy the data so it can be accessed later.

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -426,25 +570,60 @@ private void ReadExtendedAttributesBlock(Stream archiveStream)
+
+            // It is not expected that the extended attributes data section will be longer than int.MaxValue, considering
+            // 4096 is a common max path length, and also the size field is 12 bytes long, which is under int.MaxValue.
+            if (size > int.MaxValue)
```
iirc `int.MaxValue > Array.MaxLength` and since we allocate an array of given size, it would be better to check for max array size

```suggestion
            if (size > Array.MaxLength)
```

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -454,6 +633,11 @@ private void ReadGnuLongPathDataBlock(Stream archiveStream)
         {
             Debug.Assert(_typeFlag is TarEntryType.LongLink or TarEntryType.LongPath);
 
+            if (_size > int.MaxValue)
```
```suggestion
            if (_size > Array.MaxLength)
```

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -474,6 +658,31 @@ private void ReadGnuLongPathDataBlock(Stream archiveStream)
+                return null;
+            }
+
+            if (size > int.MaxValue)
```
```suggestion
            if (size > Array.MaxLength)
```

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -28,45 +30,162 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
+        // Asynchronously attempts read all the fields of the next header.
+        // Throws if end of stream is reached or if any data type conversion fails.
+        // Returns true if all the attributes were read successfully, false otherwise.
+        internal static async ValueTask<(bool, TarHeader)> TryGetNextHeaderAsync(Stream archiveStream, bool copyData, TarEntryFormat initialFormat, CancellationToken cancellationToken)
```
I currently don't have a good suggestion for how this method could be refactored to reduce code duplication. I'll try to play with the code once your PR gets merged.

carlossanlop replied:
I know. TarHeader is the problem,  since it is a struct. I may have to refactor the code later to get rid of TarHeader completely,  move everything to TarEntry.

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -53,16 +75,30 @@ internal static void CopyBytes(Stream origin, Stream destination, long bytesToCo
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(minimumLength: MaxBufferLength);
```
```suggestion
            byte[] buffer = ArrayPool<byte>.Shared.Rent((int)Math.Min(MaxBufferLength, bytesToCopy));
```

adamsitnik reviewed (commented):
I've reviewed 12 out of 64 files, will get back to this PR later today after I have some more coffee.

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -30,9 +32,9 @@ internal void WriteAsV7(Stream archiveStream, Span<byte> buffer)
+            int checksum = WriteNameSpan(buffer, out _);
             checksum += WriteCommonFields(buffer, actualLength, actualEntryType);
-            WriteChecksum(checksum, buffer);
+            _checksum = WriteChecksum(checksum, buffer);
```
why the sync method writes the checksum to a field and the async method does not?

carlossanlop replied:
TarHeader is a struct. It owns the _checksum field. In the sync method, I have no problem updating the value of _checksum inside the method: the target field gets updated. But in the async method, the TarHeader is copied into the method, so the _checksum value that gets updated is the one from the copied header, not the original one.

As discussed on chat, this is a good candidate for refactoring and simplification on a later PR.

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -42,17 +44,39 @@ internal void WriteAsV7(Stream archiveStream, Span<byte> buffer)
+                await WriteDataAsync(archiveStream, _dataStream, actualLength, cancellationToken).ConfigureAwait(false);
+            }
+
+            return finalChecksum;
```
It seems that we could reduce code duplication here by moving part of the common logic to a separate method:

```cs
private long WriteCommonFieldsAndChecksumAsV7(Span<byte> buffer)
{
    long actualLength = GetTotalDataBytesToWrite();
    TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarEntryFormat.V7);

    int checksum = WriteName(buffer, out _);
    int checksum = WriteNameSpan(buffer, out _);
    checksum += WriteCommonFields(buffer, actualLength, actualEntryType);
    WriteChecksum(checksum, buffer);
    _checksum = WriteChecksum(checksum, buffer);
    
    return actualLength;
}

internal void WriteAsV7(Stream archiveStream, Span<byte> buffer)
{
    long actualLength = WriteCommonFieldsAndChecksumAsV7(buffer);

    archiveStream.Write(buffer);

    if (_dataStream != null)
    {
        WriteData(archiveStream, _dataStream, actualLength);
    }
}

internal async Task<int> WriteAsV7Async(Stream archiveStream, Memory<byte> buffer, CancellationToken cancellationToken)
{
    cancellationToken.ThrowIfCancellationRequested();

    long actualLength = WriteCommonFieldsAndChecksumAsV7(buffer.Span);

    await archiveStream.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);

    if (_dataStream != null)
    {
        await WriteDataAsync(archiveStream, _dataStream, actualLength, cancellationToken).ConfigureAwait(false);
    }

    return _checksum;
}
```

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -62,6 +86,30 @@ internal void WriteAsUstar(Stream archiveStream, Span<byte> buffer)
+            checksum += WriteCommonFields(buffer.Span, actualLength, actualEntryType);
+            checksum += WritePosixMagicAndVersion(buffer.Span);
+            checksum += WritePosixAndGnuSharedFields(buffer.Span);
+            int finalChecksum = WriteChecksum(checksum, buffer.Span);
```
similar to the method above, can we reduce the code duplication between sync and async versions?

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -163,6 +295,36 @@ internal void WriteAsGnuInternal(Stream archiveStream, Span<byte> buffer)
+            long actualLength = GetTotalDataBytesToWrite();
+            TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarEntryFormat.Gnu);
+
+            int checksum = WriteNameMemory(buffer, out _);
```
why can't  we reuse `WriteNameSpan(buffer.Span, out _)` here?

carlossanlop committed:
Make GetBasePathForcreateFromDirectory a one-liner.

carlossanlop committed:
Make internal argument verification methods conditioned to debug.

carlossanlop committed:
Use stackalloc in TryGetNextHeader instead of ArrayPool rent.

carlossanlop committed:
Use Array.MaxLength instead of int.MaxValue for array size pre-checks.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -137,6 +245,30 @@ private static TarHeader GetGnuLongMetadataHeader(TarEntryType entryType, string
+
+            longMetadataHeader._dataStream = new MemoryStream();
+            await longMetadataHeader._dataStream.WriteAsync(Encoding.UTF8.GetBytes(longText), cancellationToken).ConfigureAwait(false);
+            longMetadataHeader._dataStream.Seek(0, SeekOrigin.Begin); // Ensure it gets written into the archive from the beginning
```
No, this is correct: I first write the data stream contents, then I seek it back to the beginning. Later, the contents of this stream will be flushed into the main archive: first we write the entry's header bits, then we write the data stream contents.

carlossanlop committed:
Use Math.Min for minimumLength in ArrayPool rent in CopyBytes* methods.

carlossanlop committed:
spacing

carlossanlop committed:
Only one WriteLeftAlignedBytesAndGetChecksum method is needed. Revert changes adding a memory-based method.

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -516,6 +773,22 @@ private static int WriteLeftAlignedBytesAndGetChecksum(ReadOnlySpan<byte> bytesT
+
+            for (int i = 0, j = 0; i < destination.Length && j < bytesToWrite.Length; i++, j++)
+            {
+                destination.Span[i] = bytesToWrite.Span[j];
```
Ah, in this case I don't need the memory based method. I can remove it.

carlossanlop committed:
Remove unnecessary TarHeader.Write memory-based methods, use only span ones.

carlossanlop committed:
Unused directive.

carlossanlop commented:
I'm seeing a weird behavior in some async tests in Linux, and I was able to reproduce it locally under certain conditions.

Take for example the test `System.Formats.Tar.Tests.TarFile_ExtractToDirectoryAsync_File_Tests.Extract_Archive_File_OverwriteFalse_Async`:

```cs
[Fact]
public Task Extract_Archive_File_OverwriteFalse_Async()
{
    string sourceArchiveFileName = GetTarFilePath(CompressionMethod.Uncompressed, TestTarFormat.pax, "file");
    using TempDirectory destination = new TempDirectory();
    string filePath = Path.Join(destination.Path, "file.txt");
    File.Create(filePath).Dispose();
    return Assert.ThrowsAsync<IOException>(() => TarFile.ExtractToDirectoryAsync(sourceArchiveFileName, destination.Path, overwriteFiles: false));
}
```

I made it non-async, and it is returning the Task directly.

If I execute all the tests locally using `dotnet build /t:test System.Formats.Tar.Tests.csproj`, all the tests pass.

But if I execute that test individually using `dotnet build /t:test System.Formats.Tar.Tests.csproj /p:XUnitMethodName=System.Formats.Tar.Tests.TarFile_ExtractToDirectoryAsync_File_Tests.Extract_Archive_File_OverwriteFalse_Async`, then it fails with the following error:

```
Starting:    System.Formats.Tar.Tests (parallel test collections = on, max threads = 12)
      System.Formats.Tar.Tests.TarFile_ExtractToDirectoryAsync_File_Tests.Extract_Archive_File_OverwriteFalse_Async [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.IO.IOException)
        Actual:   (No exception was thrown)
```

If I change the method and make it async, and make it await the Task instead of returning it, then it passes:

```
[Fact]
public async Task Extract_Archive_File_OverwriteFalse_Async()
{
    ...
    await Assert.ThrowsAsync<IOException>(() => TarFile.ExtractToDirectoryAsync(sourceArchiveFileName, destination.Path, overwriteFiles: false));
}
```

Anyone knows if this is expected or is it a bug? Regardless of the answer, it seems I'll have to change the tests that return a Task so they instead await it.

stephentoub commented:
> I made it non-async, and it is returning the Task directly.

You have a race condition.  Your code:
```C#
    using TempDirectory destination = new TempDirectory();
    string filePath = Path.Join(destination.Path, "file.txt");
    File.Create(filePath).Dispose();
    return Assert.ThrowsAsync<IOException>(() => TarFile.ExtractToDirectoryAsync(sourceArchiveFileName, destination.Path, overwriteFiles: false));
```
is using a `using` on the TempDirectory.  The moment you return the task from ThrowsAsync, that destination will be disposed.  That may or may not be before your ExtractToDirectoryAsync call has completed.

carlossanlop committed:
Prevent test failure in async tests due to race condition.

carlossanlop commented:
/azp run runtime-extra-platforms

adamsitnik commented on src/libraries/System.Formats.Tar/tests/TarReader/TarReader.TarEntry.ExtractToFile.Tests.Unix.cs:
```diff
@@ -0,0 +1,46 @@
+            using TempDirectory root = new TempDirectory();
+            using MemoryStream ms = GetTarMemoryStream(CompressionMethod.Uncompressed, TestTarFormat.ustar, "specialfiles");
+
+            TarReader reader = new TarReader(ms);
+            await using (reader)
```
the new C# syntax for using does not support async using? that's a pitty

carlossanlop replied:
Yes, it does. I was initially working on VS Code and I was getting errors saying that syntax wasn't supported. So I submitted changes like this one.

But then I started using Visual Studio 2022 again, and tried using the regular using syntax, and it works. So I joined most of the await usings like this one in my last commit, but missed a few others. I'll push another commit merging the rest of the await usings with the declaration of the disposable objects.

adamsitnik commented on src/libraries/System.Formats.Tar/tests/TarReader/TarReader.TarEntry.ExtractToFile.Tests.Unix.cs:
```diff
@@ -0,0 +1,46 @@
+            {
+                string path = Path.Join(root.Path, "output");
+
+                // Block device requires elevation for writing
```
do we need to make this test conditional then? or is the `PlatformSpecific` attribute applied above enough?

carlossanlop replied:
It seems `[PlatformSpecific(TestPlatforms.AnyUnix & ~TestPlatforms.tvOS)]` is good enough.

adamsitnik commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/SubReadStream.cs:
```diff
@@ -147,6 +155,8 @@ protected async ValueTask<int> ReadAsyncCore(Memory<byte> buffer, CancellationTo
         {
             Debug.Assert(!_hasReachedEnd);
 
+            cancellationToken.ThrowIfCancellationRequested();
+
```
> I was under the impression that the token cancellation status should be checked on every method

As long as some work was executed between them (so it could have actually changed in the meantime). At least this is how I see it.

adamsitnik reviewed (approved):
LGTM, great work Carlos!

I found some issues, but they are related to style (nits) and code duplication, not actual bugs. So please fell free to merge the PR now and we can work on the refactor later.

The amount of tests is insane! I'll be dreaming of fact and theories after reading source code of so many tests ;)

bartonjs commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarReader.cs:
```diff
@@ -198,6 +270,60 @@ private void Dispose(bool disposing)
         }
 
+        // Asynchronously disposes the current instance.
+        // If 'disposing' is 'false', the method was called from the finalizer.
+        private async ValueTask DisposeAsync(bool disposing)
```
Since the type is `sealed` it makes sense to do all of the DisposeAsync logic in DisposeAsync, instead of `DisposeAsyncCore()` (which this method should have been called for the basic async dispose pattern).

Similarly, `Dispose(bool)` can be merged into `Dispose()`, and all of the `GC.SuppressFinalize(this)`s can be removed.

carlossanlop commented on src/libraries/System.Formats.Tar/tests/CompressedTar.Async.Tests.cs:
```diff
@@ -0,0 +1,126 @@
+        [Fact]
+        public async Task TarGz_TarWriter_TarReader_Async()
+        {
+            using (TempDirectory root = new TempDirectory())
```
I decided to get rid of any potential race conditions after I found some tests to intermittently fail (asked [here](https://github.com/dotnet/runtime/pull/70574#issuecomment-1170469797), answered [here](https://github.com/dotnet/runtime/pull/70574#issuecomment-1170472272)).
